### PR TITLE
Redo of "Add ability to add output prefixes to the REPL output and use that to implement an IPython mode"

### PR DESCRIPTION
**Pull request by KristofferC**
Original PR (https://github.com/JuliaLang/julia/pull/46474) had to be reverted due to a change in a signature that happened after CI ran on that PR.

### CODE REVIEW

File: stdlib/REPL/src/REPL.jl [hunk 1]
```diff
@@ -1388,4 +1392,62 @@ function run_frontend(repl::StreamREPL, backend::REPLBackendRef)
...
+
+function out_transform(x, repl=Base.active_repl)
+    return quote
+        julia_prompt = $repl.interface.modes[1]
+        n = $repl_eval_counter(julia_prompt.hist)
+        mod = $REPL.active_module()
+        if !isdefined(mod, :Out)
+            setglobal!(mod, :Out, Dict{Int, Any}())
+        end
+        local __temp_val = $x # workaround https://github.com/JuliaLang/julia/issues/464511
```
Comment (rfourquet) [inline stdlib/REPL/src/REPL.jl:1413]: :eyes:

Comment (KristofferC) [reply stdlib/REPL/src/REPL.jl:1413]: Living in the future heh

Should be https://github.com/JuliaLang/julia/issues/46451

File: stdlib/REPL/docs/src/index.md [hunk 2]
```diff
@@ -616,6 +616,20 @@ julia> REPL.activate(CustomMod)
...
 
+## IPython mode
+
+It is possible to get an interface which is similar to the IPython REPL with numbered input prompts and output prefixes. This is done by calling `REPL.ipython_mode()`. If you want to have this enabled on startup, add
+```julia
+atreplinit() do repl
+    if !isdefined(repl, :interface)
+        repl.interface = REPL.setup_interface(repl)
+    end
+    REPL.ipython_mode(repl)
```
Comment (rfourquet) [inline stdlib/REPL/docs/src/index.md:627]: Should this be `REPL.ipython_mode!(repl)` ?

Comment (KristofferC) [reply stdlib/REPL/docs/src/index.md:627]: I guess, yeah

**Comment by PallHaraldsson** [comment]
Is it confusing (for the future) to call it an "IPython mode"?

Should it be "Jupyter mode" and module? That's what used in Julia, and IPython implementation detail, or is it even still used (or will forever)? If Jupyter is a better term, then change before 1.9 get released...

**Comment by KristofferC** [comment]
> Should it be "Jupyter mode" 

Why would it be called that?

**Comment by PallHaraldsson** [comment]
I'm not sure, I know about IPython and the history sort of. I thought Jupyter replaced IPython practically, for notebook users that would be what they know.

If this is only about numbered mode then maybe call it "Pluto mode" or just "numbered mode"?

**Comment by StefanKarpinski** [comment]
IPython still exists but is now a Python-specific REPL. This makes the Julia REPL a bit more like IPython.

**Comment by KristofferC** [comment]
>  I thought Jupyter replaced IPython practically,

https://github.com/ipython/ipython

> If this is only about numbered mode then maybe call it "Pluto mode" or just "numbered mode"?

It is not, it is about being like IPython, hence "IPython mode" :).

**Pull request merged**